---
title: 'GLDS-38 Analysis Using iDEP 0.91'
author: Hyun-seok 'Ren' Chang 
date: Fri Jul 24 13:30:41 2020
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=6, fig.height=5, fig.align = 'center') 
``` 
Analysis of [GLDS-38](https://genelab-data.ndc.nasa.gov/genelab/accession/GLDS-38/) from NASA GeneLab 

This R markdown file was auto-generated by the [iDEP website](http://ge-lab.org/idep/)
Using iDEP 0.91, originally by Steven Xijin.Ge@sdstate.edu 

Ge SX, Son EW, Yao R: iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics 2018, 19(1):534. PMID:30567491	

## 1. Read data  
First we set up the working directory to where the files are saved.    
```{r, message=FALSE} 
 setwd('~/Documents/iDEP/iDEP-GLDS/GLDS38')
 input_biclustMethod_ <- 'BCCC()' 
```
R packages used 

```{r, message=FALSE}
library(RSQLite,verbose=FALSE)	# for database connection
library(gplots,verbose=FALSE)		# for hierarchical clustering
library(ggplot2,verbose=FALSE)	# graphics
library(e1071,verbose=FALSE) 		# computing kurtosis
library(DT,verbose=FALSE) 		# for renderDataTable
library(plotly,verbose=FALSE) 	# for interactive heatmap
library(reshape2,verbose=FALSE) 	# for melt correlation matrix in heatmap

# From Data Read Function
library(edgeR,verbose=FALSE) # count data D.E. 
library(DESeq2,verbose=FALSE) # count data analysis, DEG.DESeq2

# TSNE Plot, tSNEgenePlot 
library(Rtsne,verbose=FALSE)

# PGSA Pathway PGSEA Pathway, PGSEAplot
library(PGSEA,verbose=FALSE)

# DEG.limma
library(limma,verbose=FALSE) # Differential expression
library(statmod,verbose=FALSE)

# enrichment plot
library(dendextend) # customizing tree

# enrich.net2, moduleNetwork
library(igraph)

# Stringdb_geneList, StringDB_GO_enrichmentData, stringDB_network1
# StringDB_network_link
library(STRINGdb,verbose=FALSE)

# gagePathwayData
library(gage,verbose=FALSE) # pathway analysis

#fgseaPathwayData
library(fgsea,verbose=FALSE) # fast GSEA

#ReactomePAPathwayData
library(ReactomePA,verbose=FALSE) # pathway analysis

# KeggImage
library(pathview)

# genomePlot, genomePlotDataPre
library(PREDA,verbose=FALSE)  # showing expression on genome
library(PREDAsampledata,verbose=FALSE) 
library(hgu133plus2.db,verbose=FALSE)

# biclustering
library(biclust,verbose=FALSE)

library(knitr)   #  install if needed. for showing tables with kable
library(kableExtra)

if(input_biclustMethod_ == "BCQU()" )
	library(QUBIC,verbose=FALSE) # have trouble installing on Linux
if(input_biclustMethod_ == "BCUnibic()" )
	library(runibic,verbose=FALSE) # Test biclustMethod dependant qubic runibic

# wgcna
library(WGCNA)
library(flashClust,verbose=FALSE)
```

```{r, message=FALSE  }  
 if(file.exists('iDEP_core_functions.R'))
	source('iDEP_core_functions.R') else 
    source('https://raw.githubusercontent.com/iDEP-SDSU/idep/master/shinyapps/idep/iDEP_core_functions.R') 
```  
```{r, message=FALSE}
#Each row of this matrix represents a color scheme;
mycolors_ = sort(rainbow(20))[c(1,20,10,11,2,19,3,12,4,13,5,14,6,15,7,16,8,17,9,18)]

hmcols_ <- colorRampPalette(rev(c("#D73027", "#FC8D59", "#FEE090", "#FFFFBF",
"#E0F3F8", "#91BFDB", "#4575B4")))(75)
heatColors_ = rbind(      greenred(75),     bluered(75),     colorpanel(75,"green","black","magenta"),colorpanel(75,"blue","yellow","red"),hmcols_ )
rownames(heatColors_) = c("Green-Black-Red","Blue-White-Red","Green-Black-Magenta","Blue-Yellow-Red","Blue-white-brown")
```
We are using the downloaded gene expression file where gene IDs has 
	been converted to Ensembl gene IDs. This is because the ID conversion database is too large
	to download. You can use your original file if your file uses Ensembl ID, or you do not want 
	to use the pathway files available in iDEP (or it is not available).  

```{r, message=FALSE  }  
 inputFile_ <- 'GLDS38_Expression.csv'
 sampleInfoFile_ <- 'GLDS38_Sampleinfo.csv' 
 gldsMetadataFile_ <- 'GLDS38_Metadata.csv'
 geneInfoFile_ <- 'Arabidopsis_thaliana__athaliana_eg_gene_GeneInfo.csv' #Gene symbols, location etc. 
 geneSetFile_ <- 'Arabidopsis_thaliana__athaliana_eg_gene.db'  # pathway database in SQL; can be GMT format 
 STRING10_speciesFile_ <- 'https://raw.githubusercontent.com/iDEP-SDSU/idep/master/shinyapps/idep/STRING10_species.csv' 
``` 
Parameters for reading data  
```{r, message=FALSE  } 
 input_missingValue_ <- 'geneMedian'	#Missing values imputation method
 input_dataFileFormat_ <- 1	#1- read counts, 2 FKPM/RPKM or DNA microarray (Brake up function?)
 input_minCounts_ <- 0.5	#Min counts
 input_NminSamples_ <- 1	#Minimum number of samples 
 input_countsLogStart_ <- 4	#Pseudo count for log CPM
 input_CountsTransform_ <- 1	#Methods for data transformation of counts. 1-EdgeR's logCPM 2-VST, 3-rlog 
```
```{r, message=FALSE  }  

readMetadata.out_ <- readMetadata(gldsMetadataFile_)

kable( readMetadata.out_ ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%")
```

```{r, message=FALSE  }  
 readData.out_ <- readData(inputFile_, input_missingValue=input_missingValue_, input_dataFileFormat = input_dataFileFormat_, input_minCounts=input_minCounts_, input_NminSamples = input_NminSamples_, input_countsLogStart=input_countsLogStart_, input_CountsTransform = input_CountsTransform_) 
  
   kable( head(readData.out_$data) ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```
```{r, message=FALSE  }  
 readSampleInfo.out_ <- readSampleInfo(sampleInfoFile_, readData.out_) 
 kable( readSampleInfo.out_ ) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```
```{r, message=FALSE  }  
 input_selectOrg_ ="NEW" 
 input_selectGO_ <- 'GOBP'	#Gene set category 
 input_noIDConversion_ = TRUE  
 allGeneInfo.out_ <- geneInfo(geneInfoFile_) 
 converted.out_ = NULL 
 convertedData.out <- convertedData(converted.out=converted.out_, readData.out=readData.out_, input_noIDConversion=input_noIDConversion_)	 
 nGenesFilter(readData.out=readData.out_, converted.out=converted.out_, convertedData.out=convertedData.out_, input_noIDConversion=input_noIDConversion_)  
 convertedCounts.out_ <- convertedCounts(readData.out=readData.out_, converted.out=converted.out_)  # converted counts, just for compatibility 
```

## 2. Pre-process 
```{r, message=FALSE  }  
# Read counts per library 
 parDefault = par() 
 par(mar=c(12,4,2,2)) 
 # barplot of total read counts
 x <- readData.out_$rawCounts
 groups = as.factor( detectGroups(colnames(x ) ) )
 if(nlevels(groups)<=1 | nlevels(groups) >20 )  
  col1 = 'green'  else
  col1 = rainbow(nlevels(groups))[ groups ]				
		 
 barplot( colSums(x)/1e6, 
		col=col1,las=3, main="Total read counts (millions)")  
 readCountsBias(readData.out=readData.out_, readSampleInfo.out=readSampleInfo.out_)  # detecting bias in sequencing depth 
```
```{r, message=FALSE  }  

 # Box plot 
 x = readData.out_$data 
 boxplot(x, las = 2, col=col1,
    ylab='Transformed expression levels',
    main='Distribution of transformed data') 
```
```{r, message=FALSE  }  

 #Density plot 
 par(parDefault) 
 densityPlot(readData.out = readData.out_, mycolors_)       
```
```{r, message=FALSE  }  

 # Scatter plot of the first two samples 
 plot(x[,1:2],xlab=colnames(x)[1],ylab=colnames(x)[2], 
    main='Scatter plot of first two samples') 
```
```{r, message=FALSE  }  

 ####plot gene or gene family
 input_selectOrg_ ="BestMatch" 
 input_geneSearch_ <- 'HOXA'	#Gene ID for searching 
 genePlot(allGeneInfo.out_, convertedData.out=convertedData.out_, input_selectOrg=input_selectOrg_, input_geneSearch=input_geneSearch_)  
```
```{r, message=FALSE  } 
 input_useSD_ <- 'FALSE'	#Use standard deviation instead of standard error in error bar? 
 geneBarPlotError(convertedData.out_, allGeneInfo.out_, input_selectOrg_, input_geneSearch_, input_useSD_)       
```

## 3. Heatmap  
```{r, message=FALSE  }  
 # hierarchical clustering tree
 x <- readData.out_$data
 maxGene <- apply(x,1,max)
 # remove bottom 25% lowly expressed genes, which inflate the PPC
 x <- x[which(maxGene > quantile(maxGene)[1] ) ,] 
 plot(as.dendrogram(hclust2( dist2(t(x)))), ylab="1 - Pearson C.C.", type = "rectangle") 
```
```{r, message=FALSE  }  
 #Correlation matrix
 input_labelPCC_ <- TRUE	#Show correlation coefficient? 
 correlationMatrix(readData.out=readData.out_, input_labelPCC = input_labelPCC_) 
```
```{r, message=FALSE  }  

 # Parameters for heatmap
 input_nGenes_ <- 1000	#Top genes for heatmap
 input_geneCentering_ <- TRUE	#centering genes ?
 input_sampleCentering_ <- FALSE	#Center by sample?
 input_geneNormalize_ <- FALSE	#Normalize by gene?
 input_sampleNormalize_ <- FALSE	#Normalize by sample?
 input_noSampleClustering_ <- FALSE	#Use original sample order
 input_heatmapCutoff_ <- 4	#Remove outliers beyond number of SDs 
 input_distFunctions_ <- 1	#which distant funciton to use
 input_hclustFunctions_ <- 1	#Linkage type
 input_heatColors1_ <- 1	#Colors
 input_selectFactorsHeatmap_ <- 'Gravity'	#Sample coloring factors 
 png('heatmap.png', width = 10, height = 15, units = 'in', res = 300) 
 staticHeatmap(readData.out_, readSampleInfo.out_, heatColors_, input_nGenes_, input_geneCentering_, input_sampleCentering_, input_geneNormalize_, input_sampleNormalize_, input_noSampleClustering=input_noSampleClustering_, input_heatmapCutoff=input_heatmapCutoff_, input_distFunctions=input_distFunctions_, input_hclustFunctions=input_hclustFunctions_, input_heatColors1=input_heatColors1_, input_selectFactorsHeatmap=input_selectFactorsHeatmap_) 

 dev.off()  
```
  ![heatmap] (heatmap.png)   
```{r, message=FALSE  }  
 heatmapPlotly(convertedData.out_, heatColors_, allGeneInfo.out=allGeneInfo.out_, input_geneCentering=input_geneCentering_, input_sampleCentering = input_sampleCentering_, input_geneNormalize = input_geneNormalize_, input_sampleNormalize = input_sampleNormalize_, input_heatColors1=input_heatColors1_) # interactive heatmap using Plotly 
```

## 4. K-means clustering   
```{r, message=FALSE  } 
 input_nGenesKNN_ <- 2000	#Number of genes fro k-Means
 input_nClusters_ <- 4	#Number of clusters 
 maxGeneClustering_ = 12000
 input_kmeansNormalization_ <- 'geneMean'	#Normalization
 input_KmeansReRun_ <- 0	#Random seed 

 distributionSD(convertedData.out_, input_nGenesKNN_)  #Distribution of standard deviations 
```     
```{r, message=FALSE  }  
 KmeansNclusters(convertedData.out_, input_nGenesKNN_)  #Number of clusters 
```     
```{r, message=FALSE  }  

 Kmeans.out = Kmeans(convertedData.out_, maxGeneClustering=maxGeneClustering_, input_nGenesKNN=input_nGenesKNN_, input_nClusters=input_nClusters_, input_kmeansNormalization=input_kmeansNormalization_, input_KmeansReRun=input_KmeansReRun_)   #Running K-means 
 KmeansHeatmap(.mycolors = mycolors_, .heatColors = heatColors_, .input_heatColors1=input_heatColors1_)   #Heatmap for k-Means 
 
```     
```{r, message=FALSE  }  

 #Read gene sets for enrichment analysis 
 sqlite_  <- dbDriver('SQLite') #sqlite changed to default in most functions
 input_selectGO3_ <- 'GOBP'	#Gene set category
 input_minSetSize_ <- 15	#Min gene set size
 input_maxSetSize_ <- 2000	#Max gene set size 
 GeneSets.out <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO3_,input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  )  
 # Alternatively, users can use their own GMT files by
 #GeneSets.out <- readGMTRobust('somefile.GMT')  
 results <- KmeansGO(input_nClusters=input_nClusters_)  #Enrichment analysis for k-Means clusters	
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```     
```{r, message=FALSE  } 
 input_seedTSNE_ <- 0	#Random seed for t-SNE
 input_colorGenes_ <- TRUE	#Color genes in t-SNE plot? 
 tSNEgenePlot(input_seedTSNE=input_seedTSNE_, input_colorGenes=input_colorGenes_)  #Plot genes using t-SNE 
```

## 5. PCA and beyond   
```{r, message=FALSE  } 
 input_selectFactors_ <- 'Gravity'	#Factor coded by color
 input_selectFactors2_ <- 'Preservation.Method.'	#Factor coded by shape
 input_tsneSeed2_ <- 0	#Random seed for t-SNE 
 #PCA, MDS and t-SNE plots
 PCAplot(convertedData.out_, readSampleInfo.out_, input_selectFactors_, input_selectFactors2_)  
```     
```{r, message=FALSE  }  
 MDSplot(convertedData.out_, readSampleInfo.out_, input_selectFactors_, input_selectFactors2_) 
```     
```{r, message=FALSE  }  
 tSNEplot(convertedData.out_, readSampleInfo.out_, input_selectFactors_, input_selectFactors2_, input_tsneSeed2_)  
```     
```{r, message=FALSE  }  

 #Read gene sets for pathway analysis using PGSEA on principal components 
 input_selectGO6_ <- 'GOBP' 
 GeneSets.out_ <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO6_,input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  )  
 PCApathway(convertedData.out=convertedData.out_, GeneSets.out=GeneSets.out_) # Run PGSEA analysis 
```     
```{r, message=FALSE  }  
 cat( PCA2factor(readData.out_, readSampleInfo.out=readSampleInfo.out_) )   #The correlation between PCs with factors 
```

## 6. DEG1   
```{r, message=FALSE  } 
 input_CountsDEGMethod_ <- 3	#DESeq2= 3,limma-voom=2,limma-trend=1 
 input_limmaPval_ <- 0.1	#FDR cutoff
 input_limmaFC_ <- 2	#Fold-change cutoff
 input_selectModelComprions_ <- 'Gravity: Microgravity vs. Terrestrial'	#Selected comparisons
 input_selectFactorsModel_ <- 'Gravity'	#Selected comparisons
 input_selectInteractions_ <- NULL 	#Selected comparisons
 input_selectBlockFactorsModel_ <- NULL 	#Selected comparisons
 factorReferenceLevels.out_ <- c('Gravity:Terrestrial') 

 limma.out <- limma(convertedData.out=convertedData.out_, readSampleInfo.out=readSampleInfo.out_, input_dataFileFormat = input_dataFileFormat_, input_countsLogStart = input_countsLogStart_, convertedCounts.out=convertedCounts.out_, input_CountsDEGMethod=input_CountsDEGMethod_, input_limmaPval=input_limmaPval_, input_limmaFC=input_limmaFC_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel=input_selectFactorsModel_, input_selectInteractions=input_selectInteractions_, input_selectBlockFactorsModel=input_selectBlockFactorsModel_, factorReferenceLevels.out=factorReferenceLevels.out_)
 DEG.data.out <- DEG.data(convertedData.out_, allGeneInfo.out_)
 limma.out$comparisons 
```     
```{r, message=FALSE  }  
 input_selectComparisonsVenn_ = limma.out$comparisons[1:3] # use first three comparisons
 input_UpDownRegulated_ <- FALSE	#Split up and down regulated genes 
 vennPlot(input_selectComparisonsVenn_, input_UpDownRegulated_) # Venn diagram 
```     
```{r, message=FALSE  }  
  sigGeneStats() # number of DEGs as figure 
```     
```{r, message=FALSE  }  
  sigGeneStatsTable() # number of DEGs as table 
```

## 7. DEG2   
```{r, message=FALSE  } 
 input_selectContrast_ <- 'Microgravity-Terrestrial'	#Selected comparisons 
 selectedHeatmap.data.out <- selectedHeatmap.data(convertedData.out=convertedData.out_, readSampleInfo.out=readSampleInfo.out_, .converted.out=converted.out_, .readData.out=readData.out_, .input_noIDConversion=input_noIDConversion_, input_dataFileFormat=input_dataFileFormat_, input_CountsDEGMethod=input_CountsDEGMethod_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel=input_selectFactorsModel_, factorReferenceLevels.out=factorReferenceLevels.out_, input_selectContrast=input_selectContrast_)
 selectedHeatmap(.mycolors=mycolors_,.heatColors=heatColors_)   # heatmap for DEGs in selected comparison

 # Save gene lists and data into files
 write.csv( selectedHeatmap.data(convertedData.out=convertedData.out_,  readSampleInfo.out=readSampleInfo.out_, .converted.out=converted.out_, .readData.out=readData.out_, .input_noIDConversion=input_noIDConversion_, input_dataFileFormat = input_dataFileFormat_, input_CountsDEGMethod=input_CountsDEGMethod_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel=input_selectFactorsModel_, factorReferenceLevels.out=factorReferenceLevels.out_, input_selectContrast=input_selectContrast_)$genes, 'heatmap.data.csv') 
 write.csv(DEG.data(convertedData.out_, allGeneInfo.out_),'DEG.data.csv' )
 write(AllGeneListsGMT() ,'AllGeneListsGMT.gmt')
 
```     
```{r, message=FALSE  } 
 input_selectGO2_ <- 'GOBP'	#Gene set category 
 geneListData.out <- geneListData(allGeneInfo.out_, input_selectGO2_, input_selectOrg_, input_limmaPval_, input_limmaFC_, input_selectContrast_)  
 volcanoPlot(input_limmaPval=input_limmaPval_, input_limmaFC=input_limmaFC_, input_selectContrast=input_selectContrast_)  
```     
```{r, message=FALSE  }  
  scatterPlot(convertedData.out=convertedData.out_, readSampleInfo.out=readSampleInfo.out_, input_dataFileFormat=input_dataFileFormat_, input_CountsDEGMethod=input_CountsDEGMethod_, input_limmaPval=input_limmaPval_, input_limmaFC=input_limmaFC_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel=input_selectFactorsModel_, factorReferenceLevels.out=factorReferenceLevels.out_, input_selectContrast_)  
```     
```{r, message=FALSE  }  
  MAplot(convertedData.out=convertedData.out_, readSampleInfo.out=readSampleInfo.out_, .converted.out=converted.out_, .readData.out=readData.out_, .input_noIDConversion=input_noIDConversion_, input_dataFileFormat=input_dataFileFormat_, input_CountsDEGMethod=input_CountsDEGMethod_, input_limmaPval=input_limmaPval_, input_limmaFC=input_limmaFC_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel_, factorReferenceLevels.out_, input_selectContrast_)  
```     
```{r, message=FALSE  }  
  geneListGOTable.out <- geneListGOTable()  
 # Read pathway data again 
 GeneSets.out <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO2_,input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  ) 
 input_removeRedudantSets_ <- TRUE	#Remove highly redundant gene sets? 
 results <- geneListGO(input_removeRedudantSets_)  #Enrichment analysis
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```

STRING-db API access. 
	We need to find the taxonomy id of your species, this used by STRING.
  First we try to guess the ID based on iDEP's database. Users can also skip this step and assign NCBI taxonomy id directly by
  findTaxonomyID.out = 10090 # mouse 10090, human 9606 etc.
	
```{r, message=FALSE  }  

  
 STRING10_species_ = read.csv(STRING10_speciesFile_)  
 ix = grep('Arabidopsis thaliana', STRING10_species_$official_name ) 
 findTaxonomyID.out <- STRING10_species_[ix,1] # find taxonomyID
 findTaxonomyID.out  
 
``` 
Enrichment analysis using STRING     
```{r, message=FALSE  }  
  STRINGdb_geneList.out <- STRINGdb_geneList() #convert gene lists
 input_STRINGdbGO_ <- 'Process'	#'Process', 'Component', 'Function', 'KEGG', 'Pfam', 'InterPro' 
 results <- stringDB_GO_enrichmentData(input_STRINGdbGO_)  # enrichment using STRING	
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
``` 
PPI network retrieval and analysis    
```{r, message=FALSE  } 
 input_nGenesPPI_ <- 100	#Number of top genes for PPI retrieval and analysis 
 stringDB_network1(1, input_nGenesPPI_) #Show PPI network 
``` 
Generating interactive PPI   
```{r, message=FALSE  }  
 write(stringDB_network_link(input_nGenesPPI_), 'PPI_results.html') # write results to html file 
 browseURL('PPI_results.html') # open in browser 
```

## 8. Pathway analysis   
```{r, message=FALSE  } 
 input_selectContrast1_ <- 'Microgravity-Terrestrial'	#select Comparison 
 #input_selectContrast1 = limma.out$comparisons[3] # manually set
 input_selectGO_ <- 'GOBP'	#Gene set category 
 #input_selectGO='custom' # if custom gmt file
 input_minSetSize_ <- 15	#Min size for gene set
 input_maxSetSize_ <- 2000	#Max size for gene set 
 # Read pathway data again 
 GeneSets.out <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO_,input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  ) 
 input_pathwayPvalCutoff_ <- 0.2	#FDR cutoff
 input_nPathwayShow_ <- 30	#Top pathways to show
 input_absoluteFold_ <- FALSE	#Use absolute values of fold-change?
 input_GenePvalCutoff_ <- 1	#FDR to remove genes 

 input_pathwayMethod_ = 1  # 1  GAGE
 gagePathwayData.out <- gagePathwayData(input_minSetSize=input_minSetSize_, input_maxSetSize=input_maxSetSize_, input_selectContrast1=input_selectContrast1_, input_pathwayPvalCutoff_, input_nPathwayShow_, input_absoluteFold_, input_GenePvalCutoff_)  # pathway analysis using GAGE  
   
 results <- gagePathwayData.out  #Enrichment analysis for k-Means clusters	
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```     
```{r, message=FALSE  }  
 pathwayListData.out_ = pathwayListData(allGeneInfo.out_, input_selectOrg_, input_selectGO_, input_pathwayMethod_) 
 enrichmentPlot(pathwayListData.out_, 25  ) 
```     
```{r, message=FALSE  }  
  enrichmentNetwork(pathwayListData.out_ )  
```     
```{r, message=FALSE  }  
  enrichmentNetworkPlotly(pathwayListData.out_) 
```     
```{r, message=FALSE  }  

 input_pathwayMethod_ = 3  # 1  fgsea 
 fgseaPathwayData.out <- fgseaPathwayData(input_minSetSize=input_minSetSize_, input_maxSetSize=input_maxSetSize_, input_selectContrast1=input_selectContrast1_, input_pathwayPvalCutoff=input_pathwayPvalCutoff_, input_nPathwayShow=input_nPathwayShow_, input_GenePvalCutoff=input_GenePvalCutoff_) #Pathway analysis using fgsea 
 results_ <- fgseaPathwayData.out  #Enrichment analysis for k-Means clusters	
 results_$adj.Pval <- format( results_$adj.Pval,digits=3 )
 kable( results_, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```     
```{r, message=FALSE  }  
  pathwayListData.out_ = pathwayListData(allGeneInfo.out_, input_selectOrg=input_selectOrg_, input_selectGO=input_selectGO_, input_pathwayMethod_) 
  enrichmentPlot(pathwayListData.out_, 25, mycolors_) 
```     
```{r, message=FALSE  }  
  enrichmentNetwork(pathwayListData.out_ )  
```     
```{r, message=FALSE  }  
  enrichmentNetworkPlotly(pathwayListData.out_) 
```     
```{r, message=FALSE,fig.width=9, fig.height=8  }  
   PGSEAplot(convertedData.out=convertedData.out_, readSampleInfo.out=readSampleInfo.out_, input_selectOrg = input_selectOrg_, input_dataFileFormat=input_dataFileFormat_, input_selectGO=input_selectGO_, input_minSetSize=input_minSetSize_, input_maxSetSize=input_maxSetSize_, input_CountsDEGMethod=input_CountsDEGMethod_, input_selectModelComprions=input_selectModelComprions_, input_selectFactorsModel_, factorReferenceLevels.out=factorReferenceLevels.out_, input_selectContrast1=input_selectContrast1_, input_pathwayPvalCutoff=input_pathwayPvalCutoff_, input_nPathwayShow_) # pathway analysis using PGSEA 
```

## 9. Chromosome   
```{r, message=FALSE  } 
 input_selectContrast2_ <- 'Microgravity-Terrestrial'	#select Comparison 
 #input_selectContrast2 = limma.out$comparisons[3] # manually set
 input_limmaPvalViz_ <- 0.1	#FDR to filter genes
 input_limmaFCViz_ <- 2	#FDR to filter genes 
 genomePlotly(allGeneInfo.out_, input_selectContrast2_, input_limmaPvalViz_, input_limmaFCViz_) # shows fold-changes on the genome 
```

## 10. Biclustering   
```{r, message=FALSE  } 
 input_nGenesBiclust_ <- 1000	#Top genes for biclustering
 input_biclustMethod_ <- 'BCCC()'	#Method: 'BCCC', 'QUBIC', 'runibic' ... 
 biclustering.out = biclustering(convertedData.out_, input_nGenesBiclust_, input_biclustMethod_ )  # run analysis

 input_selectBicluster_ <- 1	#select a cluster 
 biclustHeatmap(heatColors_, input_heatColors1=input_heatColors1_, input_selectBicluster=input_selectBicluster_)   # heatmap for selected cluster 
```     
```{r, message=FALSE  } 
 input_selectGO4_ <- 'GOBP'	#Gene set category 
 # Read pathway data again 
 GeneSets.out <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO4_, input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  )  
 results <- geneListBclustGO(2, input_selectBicluster=input_selectBicluster_)  #Enrichment analysis for k-Means clusters	
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```

## 11. Co-expression network    
```{r, message=FALSE  } 
 input_mySoftPower_ <- 5	#SoftPower to cutoff
 input_nGenesNetwork_ <- 1000	#Number of top genes
 input_minModuleSize_ <- 20	#Module size minimum 
 wgcna.out = wgcna(convertedData.out_, 2000, input_mySoftPower=input_mySoftPower_, input_nGenesNetwork=input_nGenesNetwork_, input_minModuleSize=input_minModuleSize_)   # run WGCNA  
```     
```{r, message=FALSE  }  
 softPower(wgcna.out)  # soft power curve 
```     
```{r, message=FALSE  }  
  modulePlot(wgcna.out)  # plot modules  
  listWGCNA.Modules.out = listWGCNA.Modules() #modules
 
```     
```{r, message=FALSE  } 
 input_selectGO5_ <- 'GOBP'	#Gene set category 
 # Read pathway data again 
 GeneSets.out <-readGeneSets( geneSetFile_,
    convertedData.out_, input_selectGO5_,input_selectOrg_,
    c(input_minSetSize_, input_maxSetSize_)  ) 
 input_selectWGCNA.Module_ <- 'Entire network'	#Select a module
 input_topGenesNetwork_ <- 10	#SoftPower to cutoff
 input_edgeThreshold <- 0.4	#Number of top genes 
 moduleNetwork(input_noIDConversion=input_noIDConversion_, allGeneInfo.out=allGeneInfo.out_, input_selectOrg_, input_selectWGCNA.Module_, input_topGenesNetwork_, input_edgeThreshold_)	# show network of top genes in selected module
 
```     
```{r, message=FALSE  } 
 input_removeRedudantSets_ <- TRUE	#Remove redundant gene sets 
 results <- networkModuleGO(2, input_selectWGCNA.Module_)  #Enrichment analysis of selected module
 results$adj.Pval <- format( results$adj.Pval,digits=3 )
 kable( results, row.names=FALSE) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  scroll_box(width = "100%") 
```   
